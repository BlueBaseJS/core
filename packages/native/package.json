{
  "name": "bluerain-boilerplate-native",
  "version": "0.3.0",
  "description": "Hello BlueRain!",
  "author": "BlueEast Team <code@blueeast.com>",
  "private": true,
  "main": "node_modules/expo/AppEntry.js",
  "scripts": {
    "start": "exp start"
  },
  "dependencies": {
    "@blueeast/bluerain-os": "git://github.com/BlueEastCode/bluerain-os.git#feature/custom_platform",
    "expo": "19.0.1",
    "react": "16.0.0-alpha.12",
    "react-dom": "^16.0.0-alpha.12",
    "react-native": "0.46.4",
    "react-native-windows": "^0.33.0",
    "reactxp": "^0.42.0-rc.25",
    "reactxp-navigation": "1.0.5",
    "reactxp-video": "^0.2.0"
  },
  "devDependencies": {
    "@types/node": "^7.0.12",
    "@types/react": "15.0.39",
    "@types/react-native": "0.46.7",
    "@types/webpack": "^2.2.14",
    "source-map-loader": "^0.1.6"
  }
}
