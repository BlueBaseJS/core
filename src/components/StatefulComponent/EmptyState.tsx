import { BlueRainType, withBlueRain } from '../../index';
import React from 'react';

const EmptyState = (props: { bluerain: BlueRainType }) => {

	const BR = props.bluerain;

	const ComponentState = BR.Components.get('ComponentState');

	return (
		<ComponentState
			title="Empty area"
			description="No data found."
		/>
	);
};

export default withBlueRain(EmptyState);
