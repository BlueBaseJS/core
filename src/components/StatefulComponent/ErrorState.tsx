import { BlueRainType,withBlueRain } from '../../index';
import Icon from '../Icon';
import React from 'react';

const ErrorIcon = () => <Icon title="Error" />;

const ErrorState = (props: { bluerain: BlueRainType }) => {

	const BR = props.bluerain;

	const ComponentState = BR.Components.get('ComponentState');

	return (
		<ComponentState
			title="Something broke!"
			description="An unknown error has occured. Please try again later."
			image={ErrorIcon}
		/>
	);
};

export default withBlueRain(ErrorState);
