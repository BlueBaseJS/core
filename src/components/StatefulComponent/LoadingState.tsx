import { BlueRain, withBlueRain } from '../../index';
import React from 'react';

const LoadingState = (props: { bluerain: BlueRain }) => {

	const BR = props.bluerain;

	const Indicator = () => <BR.Components.ActivityIndicator color="#eeeeee" size="large" />;

	return <BR.Components.ComponentState title="Loading, Please wait." image={Indicator} />;
};

export default withBlueRain(LoadingState);
