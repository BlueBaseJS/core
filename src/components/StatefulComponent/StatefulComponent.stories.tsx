import { BlueRain, withBlueRain } from '../../index';
import React from 'react';
import storiesOf from '../../../storybook/storiesOf';

storiesOf('StatefulComponent', module)
	.add('StatefulComponent children function prop', () => {

		const Story = withBlueRain((props: { bluerain: BlueRain }) => {
			const BR = props.bluerain;

			return (
				<BR.Components.StatefulComponent>
				{() =>{
					return <BR.Components.Text>Render prop pattern</BR.Components.Text>
				}}
				</BR.Components.StatefulComponent>
			);
		});

		return <Story />;
	})

	.add('StatefulComponent children prop', () => {

		const Story = withBlueRain((props: { bluerain: BlueRain }) => {
			const BR = props.bluerain;

			return (
				<BR.Components.StatefulComponent>
					<BR.Components.Text>This this a child!</BR.Components.Text>
				</BR.Components.StatefulComponent>
			);
		});

		return <Story />;
	})

	.add('StatefulComponent throw error', () => {

		const Story = withBlueRain((props: { bluerain: BlueRain }) => {
			const BR = props.bluerain;

			return (
				<BR.Components.StatefulComponent>
					{() => {
						throw new Error('Boom!');
					}}
				</BR.Components.StatefulComponent>
			);
		});

		return <Story />;
	})


	.add('Only loading Component', () => {

		const Story = withBlueRain((props: { bluerain: BlueRain }) => {
			const BR = props.bluerain;
			const LoadingState = BR.Components.get('LoadingState');

			return <LoadingState />;
		});

		return <Story />;
	})

    .add('Only EmptyState Component', () => {

	const Story = withBlueRain((props: { bluerain: BlueRain }) => {
		const BR = props.bluerain;
		const EmptyState = BR.Components.get('EmptyState');

		return <EmptyState  />;
	});

	return <Story />;
})

.add('Only ErrorState Component', () => {

	const Story = withBlueRain((props: { bluerain: BlueRain }) => {
		const BR = props.bluerain;
		const ErrorState = BR.Components.get('ErrorState');

		return <ErrorState  />;
	});

	return <Story />;
});


// storiesOf('ComponentState', module).add('Simple', () => <div>hello </div>);
