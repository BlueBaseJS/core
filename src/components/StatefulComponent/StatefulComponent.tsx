import {
	DataObserver,
	EmptyState,
	ErrorObserver,
	LoadingState,
	WaitObserver,
} from '../../getComponent';
import {
	DataObserverChildrenProps,
	StatefulComponentProps,
	WaitObserverChildrenProps,
} from '@bluebase/components';

import React from 'react';

// tslint:disable: jsdoc-format
/**
 * # üë®‚Äçüé® StatefulComponent
 *
 * This is a swiss army knife component. Intended to be used as a single source of UI
 * state management. It shows empty, loading, error or data states based on the given props.
 *
 * ## Usage
```jsx
<StatefulComponent data={data} loading={true} delay={200} timeout={10000}>
 <Text>Content</Text>
</StatefulComponent>
```
 */
export class StatefulComponent extends React.PureComponent<StatefulComponentProps> {
	public static defaultProps: Partial<StatefulComponentProps> = {
		emptyComponent: EmptyState,
		loadingComponent: LoadingState,
		timeout: 10000,
	};

	render() {
		const {
			component: Component,
			loadingComponent: LoadingComponent,
			emptyComponent: EmptyComponent,
			errorComponent,
			children,

			// DataObserver
			isLoading,
			isEmpty,
			loading,
			data,

			// WaitObserver
			delay,
			timeout,
			onRetry,
			onTimeout,

			// ErrorObserver
			error,
			checkError,

			...other
		} = this.props;

		const rest = { data, ...other };

		return (
			<ErrorObserver {...{ error, checkError, errorComponent, rest }}>
				<DataObserver {...{ isEmpty, isLoading, loading, data, rest }}>
					{(event: DataObserverChildrenProps) => {
						if (event.loading === true) {
							return React.createElement(WaitObserver, {
								children: (props: WaitObserverChildrenProps) => <LoadingComponent {...props} />,
								delay,
								onRetry,
								onTimeout,
								timeout,
							});
						}

						if (event.empty) {
							return <EmptyComponent />;
						}

						// Render 'component' prop
						if (Component) {
							return React.createElement(Component, rest);
						}

						// 'children' as a function, 'render prop' pattern
						if (typeof children === 'function') {
							return (children as any)(rest);
						}

						// children
						return children;
					}}
				</DataObserver>
			</ErrorObserver>
		);
	}
}
