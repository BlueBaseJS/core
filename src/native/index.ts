export * from './ActivityIndicator';
export * from './Button';
export * from './Image';
export * from './Text';
export * from './View';
