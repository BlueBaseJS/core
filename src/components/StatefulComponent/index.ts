import StatefulComponent, { StatefulComponentProperties } from './StatefulComponent';

import EmptyState from './EmptyState';
import ErrorState from './ErrorState';
import LoadingState from './LoadingState';

export default StatefulComponent;
export { EmptyState, ErrorState, LoadingState };

export { StatefulComponentProperties };
