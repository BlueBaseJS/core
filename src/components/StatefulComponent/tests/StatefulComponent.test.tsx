import * as Native from '../../../native';
import React from 'react';
import { StatefulComponent } from '../StatefulComponent';
import TestRenderer from 'react-test-renderer';
import WithProvider from '../../../testing/helpers/WithProvider';

beforeEach(() => {
	jest.resetModules();
});

describe('StatefulComponent', () => {
	const StatefulComponentWithProvider = (props: any) => (
		<WithProvider>
			<StatefulComponent {...props}/>
		</WithProvider>
	);

	test(`Snapshot StatefulComponent component`, () => {
		const component = TestRenderer.create(
			<StatefulComponentWithProvider />
		);
		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	test(`Snapshot StatefulComponent component with child`, () => {
		const component = TestRenderer.create(
			<StatefulComponentWithProvider>
				<Native.Text>Hello</Native.Text>
			</StatefulComponentWithProvider>
		);
		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	test(`Snapshot StatefulComponent component with child, loading: true`, () => {
		const component = TestRenderer.create(
			<StatefulComponentWithProvider loading={true}>
				<Native.Text>Hello</Native.Text>
			</StatefulComponentWithProvider>
		);
		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	test(`Snapshot StatefulComponent component with child, isEmpty: () => true`, () => {
		const component = TestRenderer.create(
			<StatefulComponentWithProvider isEmpty={() => true} loading={true}>
				<Native.Text>Hello</Native.Text>
			</StatefulComponentWithProvider>
		);
		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
