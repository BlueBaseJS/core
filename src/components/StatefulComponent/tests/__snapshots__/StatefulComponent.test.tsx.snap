// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatefulComponent Snapshot StatefulComponent component 1`] = `
<View>
  <Text>
    Empty area
  </Text>
  <Text>
    No data found.
  </Text>
</View>
`;

exports[`StatefulComponent Snapshot StatefulComponent component with child 1`] = `
<View>
  <Text>
    Empty area
  </Text>
  <Text>
    No data found.
  </Text>
</View>
`;

exports[`StatefulComponent Snapshot StatefulComponent component with child, isEmpty: () => true 1`] = `
<View>
  <Text>
    Something broke!
  </Text>
  <Text>
    An unknown error has occured. Please try again later.
  </Text>
</View>
`;

exports[`StatefulComponent Snapshot StatefulComponent component with child, loading: true 1`] = `
<View>
  <Text>
    Something broke!
  </Text>
  <Text>
    An unknown error has occured. Please try again later.
  </Text>
</View>
`;
